### PR TITLE
feat(pipeline): Add the ability to edit all nodes in the new UI

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

This solves two major problems in the new pipeline UI:
- `Plugin` and `BatchExport` being completely separate types with little overlap, making it horribly cumbersome and repetitive to handle them in the new Pipeline product, whose philosophy is pretty much turned around from the previous Apps product (we're going from Apps + Batch Exports to Filters + Transformations + Destinations, collectively nodes)
- …which I realized acutely implementing functional configuration UI logic for Pipeline nodes - this is the actual goal of this PR

## Changes

Adds a `types.ts` file for Pipeline, with the whole structure organized in a way that makes writing logic much easier (see the file for details).

Then adds the ability to configure all existing pipeline nodes with proper validation and update handling.

## How did you test this code?

Manual QA so far.

### CODE REVIEW

**Comment by Twixes** [comment]
To make it easier to wrap my head (and our heads in general) around how the Pipeline works, I've written a short doc (though I imagine there's already been an RFC about this somewhere, this was more for collecting thoughts):

# Pipeline UX

The Pipeline product is composed of four stages:

1. **Sources** (not exposed yet)
   *Examples:* our SDKs
2. **Filters**
   *Jobs:* Adherence to privacy rules by filtering events out conditionally (e.g. Filter Out Plugin), ingestion-time sampling (e.g. Downsampling Plugin)
3. **Transformations**
   *Jobs:* Enrichment of event/person properties (e.g. GeoIP), normalization of them (e.g. PostHog Netdata Event Processing), adherence to privacy rules by filtering event properties out conditionally (e.g. Property Filter)
4. **Destinations**
   *Examples:* S3, Customer.io, Replicator
   *Jobs:* Exporting events for further analysis (e.g. BigQuery), exporting events for integration with other tooling (e.g. Customer.io)
   
Collectively, the above are called _pipeline nodes_, since they form a directed acyclic graph that describes event processing. Here's an example pipeline:

```mermaid
flowchart TD
    subgraph Sources
    style Sources stroke-dasharray: 4
    A(Python SDK)
    B(Browser SDK)
    end
    
    E(Capture endpoint)
    style E stroke-dasharray: 4
    A --> E
    B --> E
    E --> Filters

    subgraph Filters
    X(Filter X)
    Y(Filter Y)
    end
    Filters --> Transformations

    subgraph Transformations
    1(GeoIP) --> 2
    2(Normalize URLs)
    end
    Transformations --> 3
    Transformations --> 4
    subgraph Destinations
    3(Snowflake)
    4(Customer.io)
    end
```

## Ordering

Sources and Destinations are not ordered, but Filters and Transformations are. Because order can have important consequences for Transformations, it's configurable for that stage. In Filters however it isn't actually relevant, so while there's _some_ internal ordering, it's not exposed.

## Strict Filters/Transformations risk management

When it comes to Filters/Transformations, there are three levels of risk we must be mindful of:

1. PostHog-built sync logic, e.g. GeoIP
  ✅ Safe, so this will be our primary approach going forward (the caveat is that we're only able to maintain logic that's useful for a broad segment of users)
2. Custom sync logic written by users, e.g. PostHog Netdata Event Processing
  ❗️ Unsafe, UNLESS we securely isolate the full JS VMs _or_ build our own intentionally simple runtime (Marius's HogVM, that would be great for the 80% of cases which are just `if/else: <set some property>`); we can sometimes work out something special for high-paying customers, but it's a form of support, not self-serve
3. Any async logic, meaning external sources, e.g. Clearbit Enhance User
  ‼️ Unsafe, as async processing breaks guarantees important for the resiliency of ingestion
  
  Plugins falling under 3 will be phased out, and those falling under 2 will be pruned.
  
## Destinations unified

Because of the complicated history of Destinations, some of them are powered by one backend, and the rest by another:

- Webhook destinations are powered by plugins. They send events individually in real time.
- Batch export destinations are powered by Temporal. They send events in batches with specified frequency.

While these two used to be completely separate parts of the Apps product, we're unifying their experience within the Pipeline product. A destination is a destination no matter if it's real-time or batched by hour, hence the UX should be the same.

File: frontend/src/scenes/pipeline/PipelineNodeConfiguration.tsx [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import { useActions, useValues } from 'kea'
+import { Form } from 'kea-forms'
+import { Field } from 'lib/forms/Field'
+import { IconLock } from 'lib/lemon-ui/icons'
+import { LemonMarkdown } from 'lib/lemon-ui/LemonMarkdown'
+import React from 'react'
+import { BatchExportsEditFields } from 'scenes/batch_exports/BatchExportEditForm'
+import { BatchExportConfigurationForm } from 'scenes/batch_exports/batchExportEditLogic'
+import { getConfigSchemaArray, isValidField } from 'scenes/pipeline/configUtils'
+import { PluginField } from 'scenes/plugins/edit/PluginField'
```
Comment (tiina303) [inline frontend/src/scenes/pipeline/PipelineNodeConfiguration.tsx:11]: could we move the code to this side & import from the other side, would be easier to nuke later

Comment (Twixes) [reply frontend/src/scenes/pipeline/PipelineNodeConfiguration.tsx:11]: We can, but it make this PR even less readable, so I think it's worth doing that in a separate PR

File: frontend/src/scenes/pipeline/pipelineNodeLogic.tsx [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    loaders(({ props, values }) => ({
+        node: [
+            null as PipelineNode | null,
+            {
+                loadNode: async (_, breakpoint) => {
+                    if (!props.stage) {
+                        return null
+                    }
+                    let node: PipelineNode
+                    if (typeof props.id === 'string') {
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/pipeline/pipelineNodeLogic.tsx:68]: Hmm, it doesn't work with `new` YET

File: frontend/src/scenes/pipeline/types.ts
Comment (Twixes) [inline frontend/src/scenes/pipeline/types.ts:1]: This types file is the core of this PR

**Comment by benjackwhite** [comment]
I'll do a deep dive today review to get this in 👍

**Review by tiina303 — APPROVED** [review]
We should likely just merge this monster (fix the code quality 2 errors) & then we can address follow-ups in separate PRs.

Played around locally and 
✅  old UI: editing batch exports and plugins
✅  old UI: creating new batch exports and plugins
✅  3000: editing existing plugins
⛔  3000: editing batch exports doesn't work, it showed me 
This destination isn't configurable."
⛔  3000: creating new plugin nor batch exports configs doesn't work, it showed me "Transformation not found"
⛔  3000: the transformation and destination links in the breadcrumbs send me to overview page instead of respective stage
⛔  3000: can't edit the name & description of plugins nor batch exports

File: frontend/src/scenes/pipeline/Pipeline.stories.tsx [hunk 3]
```diff
@@ -23,12 +33,17 @@ export default {
...
+                '/api/organizations/:organization_id/pipeline_transformations/': plugins,
+                '/api/projects/:team_id/pipeline_transformation_configs/': pluginConfigs,
+                '/api/projects/:team_id/pipeline_transformation_configs/:id': pluginConfigRetrieveMock,
+                '/api/organizations/:organization_id/pipeline_destinations/': plugins,
                 '/api/projects/:team_id/pipeline_destination_configs/': pluginConfigs,
+                '/api/projects/:team_id/pipeline_destination_configs/:id': pluginConfigRetrieveMock,
+
                 '/api/projects/:team_id/app_metrics/:plugin_config_id?date_from=-7d': require('./__mocks__/pluginMetrics.json'),
                 '/api/projects/:team_id/app_metrics/:plugin_config_id/error_details?error_type=Error': require('./__mocks__/pluginErrorDetails.json'),
             },
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/pipeline/Pipeline.stories.tsx:49]: Should be fixed by #19993 now

File: frontend/src/scenes/batch_exports/BatchExportEditForm.tsx [hunk 4]
```diff
@@ -507,3 +64,431 @@ export function BatchExportsEditForm(props: BatchExportsEditLogicProps): JSX.Ele
...
+                        </Field>
+                    )}
+                </div>
+
+                <LemonBanner type="info">
+                    This batch exporter will schedule regular batch exports at your indicated interval until the end
+                    date. Once you have configured your exporter, you can trigger a manual export for historic data.
+                </LemonBanner>
+
+                {isNew && !isPipeline ? (
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/batch_exports/BatchExportEditForm.tsx:142]: I see the use case. The thing is that in the new UI "enabled" will be a checkbox that shows up the same for all types of nodes, both being created and edited.

File: frontend/src/scenes/pipeline/Destinations.tsx [hunk 5]
```diff
@@ -129,69 +137,56 @@ export const DestinationMoreOverlay = ({
...
+                },
+                {
+                    label: 'View metrics',
+                    to: urls.pipelineNode(PipelineStage.Destination, destination.id, PipelineNodeTab.Metrics),
+                },
+                {
+                    label: 'View logs',
+                    to: urls.pipelineNode(PipelineStage.Destination, destination.id, PipelineNodeTab.Logs),
+                },
+                // TODO: Add link to source code for staff
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/pipeline/Destinations.tsx:172]: We should add it back, but we don't yet have a way of linking to apps in the Apps Management page, so in another PR

**Pull request merged**